snuba/clickhouse/translator/rules.py

```diff
@@ -0,0 +1,85 @@
+from dataclasses import dataclass
+from typing import Optional
+
+from snuba.clickhouse.query import Expression
+from snuba.datasets.plans.translator.visitor import ExpressionMapper
+from snuba.query.dsl import array_element
+from snuba.query.expressions import (
+    Column,
+    ExpressionVisitor,
+    FunctionCall,
+    SubscriptableReference,
+)
+
+
+@dataclass(frozen=True)
+class ColumnMapper(ExpressionMapper[Column, Expression]):
+    """
+    Maps a column with a name and a table into a column with a different name and table.
+
+    The alias is not transformed.
+    """
+
+    from_col_name: str
+    from_table_name: Optional[str]
+    to_col_name: str
+    to_table_name: Optional[str]
+
+    def attemptMap(
+        self, expression: Column, children_translator: ExpressionVisitor[Expression],
+    ) -> Optional[Expression]:
+        if (
+            expression.column_name == self.from_col_name
+            and expression.table_name == self.from_table_name
+        ):
+            return Expression(
+                Column(
+                    alias=expression.alias,
+                    table_name=self.to_table_name,
+                    column_name=self.to_col_name,
+                )
+            )
+        else:
+            return None
+
+
+@dataclass(frozen=True)
+class TagMapper(ExpressionMapper[SubscriptableReference, Expression]):
+    """
+    Basic implementation of a tag mapper that transforms a subscriptable
+    into a Clickhouse array access.
+    """
+
+    from_column_name: str
+    from_column_table: Optional[str]
+    to_col_name: str
+    to_table_name: Optional[str]
+
+    def attemptMap(
```

Comment:
We discussed in the past whether tag promotion should be in the translator or in a query processor. There were advantages and drawbacks of both solutions.
The advantage of doing it in a processor is that the translator would be simpler and easier to express in a declarative way.
The advantage of doing it in the translator, is that it would be easier to identify tags expressions. This is the example of doing that in a processor: https://github.com/getsentry/snuba/pull/955/files#diff-17bcbc40c66ec928ca0b4e13622325e5R44-R59

I think that verbosity would be mitigated a lot by https://github.com/getsentry/snuba/pull/942 so I don't see a strong advantage in doing tag promotion in the translator which would then need a lot more context on the schema to do tag promotion.